streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -4615,7 +4732,110 @@ void to(final TopicNameExtractor<K, V> topicExtractor,
      * @see #foreach(ForeachAction)
      * @see #transform(TransformerSupplier, String...)
      */
-    void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
+    <KOut, VOut> void process(final ProcessorSupplier<? super K, ? super V, KOut, VOut> processorSupplier,
+                              final String... stateStoreNames);
```

Comment:
Here is the new API. Note: the forwarding type should technically be Void, since you can't forward from these processors, but I left it unbounded to potentially be compatible with whatever we decide to do in https://issues.apache.org/jira/browse/KAFKA-10603

Reply:
Just wondering, if we ended up not doing KAFKA-10603, then how to bound `KOut/Vout` to `Void` only?

Reply:
Thanks, @guozhangwang !

Yeah, it's kind of an either/or thing. If we make it `KOut, VOut` now, we can't tighten it to `Void, Void` later. 

Although, coming back to it, I'm thinking that we actually should be able to make it `Void, Void` now and then relax it to `KOut, VOut` later. Maybe you can sanity-check me:

Now:
```
Def:
void process(final ProcessorSupplier<? super K, ? super V, Void, Void> processorSupplier)

Usage:
mystream.process(new ProcessorSupplier<String, String, Void, Void>(){
...
  Processor<String, String, Void, Void> get() {
    // calling forward is optional, but if they do it, they could only call:
    context.forward(new Record(null, null, ...));
  }
...
})
```

Later:
```
Def:
KStream<KOut, VOut> process(final ProcessorSupplier<? super K, ? super V, KOut, VOut> processorSupplier)

Same Usage is still ok:
* Return is unused, which is fine
* Return KOut and VOut both bind to Void, which is also fine.
mystream.process(new ProcessorSupplier<String, String, Void, Void>(){
...
  Processor<String, String, Void, Void> get() {
    // calling forward is optional, but if they do it, they could only call:
    context.forward(new Record(null, null, ...));
  }
...
})
```

So it seems we actually could/should set them to `Void` right now. Does that seem right to you?

Reply:
If any template type `K` could bind to `Void` then that seems fine -- I was not clear about this myself :P

Reply:
@vvcephei +1 I like the last proposal: using `Void` and keep the option to migrate to `K/V` later.